pkg/server/sotw/v3/watches.go

```diff
@@ -0,0 +1,75 @@
+package sotw
+
+import (
+	"context"
+	"reflect"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+)
+
+// watches for all xDS resource types
+type watches struct {
+	responders map[string]*watch
+
+	// indexes is a list of indexes for each dynamic select case which match to a watch
+	cases []reflect.SelectCase
+}
+
+// newWatches creates and initializes watches.
+func newWatches() watches {
+	// deltaMuxedResponses needs a buffer to release go-routines populating it
+	return watches{
+		responders: make(map[string]*watch, int(types.UnknownType)),
+		cases:      make([]reflect.SelectCase, 2), // We use 2 for the default computation here: ctx.Done() + reqCh.Recv()
+	}
+}
+
+// Cancel all watches
+func (w *watches) Cancel() {
+	for _, watch := range w.responders {
+		watch.Cancel()
+	}
+}
+
+// recomputeWatches will analyze the currently typed list of the known watches and increase the known list of dynamic channels if needed
```

Comment:
I would just say rebuilds the list of dynamic channels, I think it could go both up and down